dynoid/dynoid.go

```diff
@@ -5,43 +5,234 @@ import (
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
+	"log/slog"
+	"os"
+	"path"
+	"strconv"
 	"strings"
 	"sync"
 
 	"github.com/coreos/go-oidc/v3/oidc"
 )
 
-type IssuerCallback func(string) error
+const (
+	AudienceHeroku = "heroku"
+)
 
-type Verifier struct {
-	IssuerCallback IssuerCallback
-	config         *oidc.Config
-	providers      sync.Map
+// Returned by an IssuerCallback get's an issuer it doesn't trust
+type UntrustedIssuerError struct {
+	Issuer string
 }
 
-func New(clientID string) *Verifier {
-	return &Verifier{config: &oidc.Config{ClientID: clientID}}
+func (e *UntrustedIssuerError) Error() string {
+	return fmt.Sprintf("untrusted issuer: %v", e.Issuer)
+}
+
+// Returned when the token doesn't match the expected format
+type MalformedTokenError struct {
+	Err error
 }
 
+func (e *MalformedTokenError) Error() string {
+	return fmt.Sprintf("malformed token: %s", e.Err.Error())
+}
+
+func (e *MalformedTokenError) Unwrap() error {
+	return e.Err
+}
+
+type staticError string
+
+func (e staticError) Error() string {
+	return string(e)
+}
+
+const (
+	ErrMustCheckIssuer staticError = "must check issuer"
+)
+
+// An IssuerCallback is called whenever a token is verified to ensure it matches
+// some expected criteria.
+type IssuerCallback func(issuer string) error
+
+// AllowHerokuHost verifies that the issuer is from Heroku for the given host
+// domain
 func AllowHerokuHost(host string) IssuerCallback {
 	return func(issuer string) error {
 		if !strings.HasPrefix(issuer, fmt.Sprintf("https://oidc.%v/", host)) {
-			return fmt.Errorf("untrusted issuer: %v", issuer)
+			return &UntrustedIssuerError{Issuer: issuer}
 		}
 
 		return nil
 	}
 }
 
+// Subject contains all of the subject information stored by Heroku when issuing
+// a token.
+type Subject struct {
+	AppID      string `json:"app_id"`
+	AppName    string `json:"app_name"`
+	DynoName   string `json:"dyno_name"`
+	DynoNumber int    `json:"dyno_number"`
+}
+
+func (s *Subject) LogValue() slog.Value {
+	return slog.GroupValue(
+		slog.String("app_id", s.AppID),
+		slog.String("app_name", s.AppName),
+		slog.String("dyno", fmt.Sprintf("%s.%d", s.DynoName, s.DynoNumber)),
+	)
+}
+
+func (s *Subject) MarshalText() ([]byte, error) {
+	return []byte(s.String()), nil
+}
+
+func (s *Subject) UnmarshalText(text []byte) error {
+	if s == nil {
+		*s = Subject{}
+	}
+
+	sub := string(text)
+	parts := strings.Split(sub, ":")
+	if len(parts) != 5 || parts[0] != "app" || parts[3] != "dyno" {
+		return fmt.Errorf("unexpected subject format: %q", sub)
+	}
+
+	app := strings.Split(parts[1], ".")
+	dyno := strings.Split(parts[4], ".")
+
+	if len(app) != 2 || len(dyno) != 2 {
+		return fmt.Errorf("unexpected subject format: %q", sub)
+	}
+
+	s.AppID = app[0]
+	s.AppName = app[1]
+	s.DynoName = dyno[0]
+	s.DynoNumber, _ = strconv.Atoi(dyno[1])
```

Comment:
This is a good example of where we need a later change due to [DR: Deprecating Dyno Numbering on the Heroku Platform](https://salesforce.quip.com/7teIAUvPUQiG) and [RFC: Fir Dyno Naming Separator](https://salesforce.quip.com/SiDbAIjz5w2i).

(Pinging @dianadoherty and @ricardochimal for their awareness.)

Reply:
I was thinking about this after watching @dianadoherty's VAT presentation this morning. Thanks for confirming...

What do you think about just returning `s.Dyno` and not parsing out the name from the number?